### PR TITLE
REFACTOR-#4631: Add mypy checks for modin.distributed

### LINKED ISSUE #4631: Add mypy checks for modin.distributed
We need to expand our mypy checks to cover Python files in `modin/distributed/`. There may be a lot of files to modify at once so folks should focus on type annotations one file at a time. Please make sure to add files to `mypy.ini`!

### PR DESCRIPTION
**Pull request by rosdyana**
Signed-off-by: Rosdyana Kusuma <rosdyana.kusuma@gmail.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4631 
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/distributed/dataframe/pandas/partitions.py [hunk 1]
```diff
@@ -141,10 +147,16 @@ def from_partitions(
     from modin.core.execution.dispatching.factories.dispatcher import FactoryDispatcher
 
     factory = FactoryDispatcher.get_factory()
-
-    partition_class = factory.io_cls.frame_cls._partition_mgr_cls._partition_class
+    assert factory is not None
+    assert factory.io_cls is not None
+    assert factory.io_cls.frame_cls is not None
+    assert factory.io_cls.frame_cls._partition_mgr_cls is not None
```
Comment (mvashishtha) [inline modin/distributed/dataframe/pandas/partitions.py:153]: These shouldn't be `None`. Does mypy complain that they could be `None`?

**Review by mvashishtha — COMMENTED** [review]
@rosdyana thank you so much for making this PR. I have left some suggestions inline.

There will likely be some errors once you accept my suggestion to edit the mypy.ini file so it includes `modin/distributed/dataframe/pandas/partitions.py`. Please try to address them and check your work locally by running `mypy --config-file mypy.ini modin/distributed`. If you have any questions or you get stock, please comment here.

File: modin/distributed/dataframe/pandas/partitions.py (hunk 1 above)
Comment (rosdyana) [reply modin/distributed/dataframe/pandas/partitions.py:153]: Hi @mvashishtha this is the complaints from mypy
```
modin/distributed/dataframe/pandas/partitions.py:157:27: error: "None" has no attribute "_partition_mgr_cls"  [attr-defined]

modin/distributed/dataframe/pandas/partitions.py:158:23: error: "None" has no attribute "_partition_mgr_cls"  [attr-defined]
```
Do you think we can ignore it?

File: mypy.ini [hunk 2]
```diff
@@ -21,4 +21,5 @@ warn_unreachable=True
 files = 
         modin/config/*.py,
         modin/logging/*.py,
+        modin/distributed/*.py,
         modin/*.py
```
[the remark this answers is not in the document]
Comment (rosdyana) [reply mypy.ini]: Fixed

File: modin/distributed/dataframe/pandas/partitions.py [hunk 3]
```diff
@@ -13,13 +13,14 @@
...
 
+from typing import Optional
 import numpy as np
 
 from modin.core.storage_formats.pandas.query_compiler import PandasQueryCompiler
 from modin.pandas.dataframe import DataFrame
 
 
-def unwrap_partitions(api_layer_object, axis=None, get_ip=False):
+def unwrap_partitions(api_layer_object: DataFrame, axis: int = None, get_ip: bool = False) -> list:
```
[the remark this answers is not in the document]
Comment (rosdyana) [reply modin/distributed/dataframe/pandas/partitions.py]: Fixed

File: modin/distributed/dataframe/pandas/partitions.py [hunk 4]
```diff
@@ -99,8 +100,13 @@ def get_block(partition):
...
 
 def from_partitions(
-    partitions, axis, index=None, columns=None, row_lengths=None, column_widths=None
-):
+    partitions: list,
+    axis: int,
+    index: Optional[list] | None = None,
+    columns: Optional[list] | None = None,
+    row_lengths: Optional[list] | None = None,
+    column_widths: Optional[list] | None = None
```
[the remark this answers is not in the document]
Comment (rosdyana) [reply modin/distributed/dataframe/pandas/partitions.py]: Fixed

File: modin/distributed/dataframe/pandas/partitions.py [hunk 5]
```diff
@@ -54,10 +55,10 @@ def unwrap_partitions(api_layer_object, axis=None, get_ip=False):
     modin_frame._propagate_index_objs(None)
     if axis is None:
 
-        def _unwrap_partitions():
+        def _unwrap_partitions() -> list:
             [p.drain_call_queue() for p in modin_frame._partitions.flatten()]
 
-            def get_block(partition):
+            def get_block(partition: DataFrame) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (rosdyana) [reply modin/distributed/dataframe/pandas/partitions.py]: Fixed

File: modin/distributed/dataframe/pandas/partitions.py (hunk 1 above)
Comment (pyrito) [reply modin/distributed/dataframe/pandas/partitions.py]: It may be that we need to add mypy hints to the FactoryDispatcher.

File: modin/distributed/dataframe/pandas/partitions.py [hunk 6]
```diff
@@ -54,10 +61,10 @@ def unwrap_partitions(api_layer_object, axis=None, get_ip=False):
     modin_frame._propagate_index_objs(None)
     if axis is None:
 
-        def _unwrap_partitions():
+        def _unwrap_partitions() -> list:
             [p.drain_call_queue() for p in modin_frame._partitions.flatten()]
 
-            def get_block(partition):
+            def get_block(partition: Union[PandasOnRayDataframePartition, PandasOnDaskDataframePartition]) -> np.ndarray:
```
Comment (dchigarev) [inline modin/distributed/dataframe/pandas/partitions.py:67]:
```suggestion
            def get_block(partition: PandasDataframePartition) -> np.ndarray:
```
maybe it's worth using the base class here for simplicity? Judging by the function's implementation it definitely can process  the `PandasDataframePartition` class itself as well as any of its descendants

**Comment by mvashishtha** [comment]
@rosdyana in order to pass the commit lint, you will have to amend your first commit so that it says something like, "FIX-#4631: Add mypy checks for modin.distributed."

You should be able to amend your first commit with:

1. `git rebase -i d5bd6376d5b7735c4c676e74bd73ef36edcdfdff~1`
2. You should then get an editable list of actions for each commit. Choose "r" for reword for your first commit, then "p" for each of the other commits to keep them on top of the first one.
Screen Shot 2022-10-10 at 10 42 11 AM
3. Save and exit the editable list of actions.
4. You should get an editor again asking you to reword your initial commit. Do so.
Screen Shot 2022-10-10 at 10 51 14 AM
5. Resolve merge conflicts

File: modin/distributed/dataframe/pandas/partitions.py (hunk 6 above)
Comment (pyrito) [reply modin/distributed/dataframe/pandas/partitions.py:67]: Yes we should most definitely use the base class here.

Comment (mvashishtha) [reply modin/distributed/dataframe/pandas/partitions.py:67]: If you look further down in this function, the function only works with these two instances. I don't see why other subclasses of `PandasDataframePartition` wouldn't work, but I wanted to keep the typing here consistent with what would really work in this function. I don't have a strong opinion either way.

Comment (rosdyana) [reply modin/distributed/dataframe/pandas/partitions.py:67]: I agree with @mvashishtha.

Comment (noloerino) [reply modin/distributed/dataframe/pandas/partitions.py:67]: Actually, what's the reason for that assertion? I don't see any part in the resulting control flow that requires the partitions to be specifically either PandasOnRay or PandasOnDask variants.

**Comment by mvashishtha** [comment]
@rosdyana it looks like you've added commit e90473655bdff4d082b378db4145c95ad92e5509 with the correct formatting, but your first commit still has the incorrect formatting. Your commits as they are will still fail the CI. Could you please try rewording your **first** commit?

Meanwhile, I'll respond to the open inline comments. Please ping this thread if I haven't responded within 24 hours from now.

**Comment by mvashishtha** [comment]
@rosdyana just FYI, next time you want to accept someone's suggestion on a GitHub PR, I highly recommend using the "commit suggestion" if you want to accept one suggestion, or the "add suggestion to batch" button if you have multiple suggestions. That usually reduces the amount of work you have to do and makes it easier for everyone to track what's happening with the PR.

Screen Shot 2022-10-11 at 12 10 05 AM

**Comment by rosdyana** [comment]
> @rosdyana in order to pass the commit lint, you will have to amend your first commit so that it says something like, "FIX-#4631: Add mypy checks for modin.distributed."
> 
> You should be able to amend your first commit with:
> 
> 1. `git rebase -i d5bd6376d5b7735c4c676e74bd73ef36edcdfdff~1`
> 2. You should then get an editable list of actions for each commit. Choose "r" for reword for your first commit, then "p" for each of the other commits to keep them on top of the first one.
>    Screen Shot 2022-10-10 at 10 42 11 AM
> 3. Save and exit the editable list of actions.
> 4. You should get an editor again asking you to reword your initial commit. Do so.
>    Screen Shot 2022-10-10 at 10 51 14 AM
> 5. Resolve merge conflicts

I tried to amend that commit, but it doesn't work, I have no idea. Do you mind if I re-create the PR?

**Comment by mvashishtha** [comment]
@rosdyana what's the error that you got? Did you try to "reword" the first commit and "squash" or "pick" the other commits? Rebasing can be hard, but it's better for history keeping and easier for everyone involved with the PR if we can keep all the comment history in a single PR.

**Comment by rosdyana** [comment]
> @rosdyana what's the error that you got? Did you try to "reword" the first commit and "squash" or "pick" the other commits? Rebasing can be hard, but it's better for history keeping and easier for everyone involved with the PR if we can keep all the comment history in a single PR.

Yes, I did "reword" for the first commit, and then choose "pick" for the other commits. Should I "squash" the other commits?

**Comment by mvashishtha** [comment]
@rosdyana I'm going to remove the commit requirement in #4982. Please don't worry about the commit message or rebasing at all for now.

File: modin/distributed/dataframe/pandas/partitions.py (hunk 6 above)
Comment (mvashishtha) [reply modin/distributed/dataframe/pandas/partitions.py:67]: @noloerino this method has to return ray or dask futures, so it can only work with ray or dask virtual or block partitions. The virtual partitions can have multiple blocks but this method assumes there's only one block. So only the ray and dask block partitions work.

@dchigarev the assertion prevents this function from working with `PandasDataframePartition` descendant `PandasOnPythonDataframePartition`, but that aside, the docstring says it should only return ray and dask futures.

Going back to the `get_block` code in question in this PR, I still prefer keeping it as is to match the assertion. We might be able to make the list less complicated if we generalize this function to include other partition types.

File: modin/distributed/dataframe/pandas/partitions.py [hunk 7]
```diff
@@ -13,13 +13,20 @@
...
 
 from modin.core.storage_formats.pandas.query_compiler import PandasQueryCompiler
-from modin.pandas.dataframe import DataFrame
+from modin.pandas.dataframe import DataFrame, Series
+from modin.core.execution.ray.implementations.pandas_on_ray.partitioning import (
+    PandasOnRayDataframePartition,
+)
+from modin.core.execution.dask.implementations.pandas_on_dask.partitioning.partition import (
+    PandasOnDaskDataframePartition,
+)
```
Comment (vnlitvinov) [inline modin/distributed/dataframe/pandas/partitions.py:26]: this should be done under `if TYPING` or similar condition, as, if user doesn't have one of these installed, they won't be able to use `modin.distributed` even if they should be able to

File: modin/distributed/dataframe/pandas/partitions.py (hunk 6 above)
Comment (vnlitvinov) [reply modin/distributed/dataframe/pandas/partitions.py:67]: For this PR, we should keep the typing as close to actual behaviour as possible. I think we should refactor the function so that it works with any descendant of `PandasDataframePartition` (and shorten the typing) in a separate PR.

File: modin/distributed/dataframe/pandas/partitions.py [hunk 8]
```diff
@@ -99,8 +106,13 @@ def get_block(partition):
 
 
 def from_partitions(
-    partitions, axis, index=None, columns=None, row_lengths=None, column_widths=None
-):
+    partitions: list,
+    axis: Optional[int],
+    index: Optional[list] = None,
+    columns: Optional[list] = None,
```
Comment (vnlitvinov) [inline modin/distributed/dataframe/pandas/partitions.py:112]: This `list` doesn't feel correct, `index` and `columns` could easily be descendants of `pandas.Index` or `ndarray`, maybe we better specify them as `Optional[Sequence]` instead?

File: modin/distributed/dataframe/pandas/partitions.py [hunk 9]
```diff
@@ -141,10 +153,9 @@ def from_partitions(
     from modin.core.execution.dispatching.factories.dispatcher import FactoryDispatcher
 
     factory = FactoryDispatcher.get_factory()
-
-    partition_class = factory.io_cls.frame_cls._partition_mgr_cls._partition_class
     partition_frame_class = factory.io_cls.frame_cls
     partition_mgr_class = factory.io_cls.frame_cls._partition_mgr_cls
+    partition_class = factory.io_cls.frame_cls._partition_mgr_cls._partition_class
```
Comment (vnlitvinov) [inline modin/distributed/dataframe/pandas/partitions.py:158]: what's the reason for the change

File: modin/distributed/dataframe/pandas/partitions.py (hunk 1 above)
Comment (mvashishtha) [reply modin/distributed/dataframe/pandas/partitions.py]: @rosdyana could you please add the assertions back so we stop getting these failures, and also add a comment like this:

# TODO(https://github.com/modin-project/modin/issues/5127): Remove these assertions once the dependencies of this function all have types.

**Review by mvashishtha — COMMENTED** [review]
@rosdyana I left some more comments. Please respond here or in the comment threads if you have any questions or concerns. Also, please be sure to run `mypy --config-file=mypy.ini modin/distributed/` locally from the modin root to check whether mypy is passing. If it fails and you can't fix the failures, just say so here and we'll try to come to a resolution.

Once again, there's no need to rebase or amend your first commit.

**Comment by rosdyana** [comment]
> @rosdyana I left some more comments. Please respond here or in the comment threads if you have any questions or concerns. Also, please be sure to run `mypy --config-file=mypy.ini modin/distributed/` locally from the modin root to check whether mypy is passing. If it fails and you can't fix the failures, just say so here and we'll try to come to a resolution.
> 
> Once again, there's no need to rebase or amend your first commit.

Hi @mvashishtha the mypy returns this error
```
modin/distributed/dataframe/pandas/partitions.py:168:5: error: Statement is unreachable  [unreachable]
Found 1 error in 1 file (checked 4 source files)
```
https://github.com/modin-project/modin/blob/88ef64c2ceeb921dd151906c31dba84517eddb21/modin/distributed/dataframe/pandas/partitions.py#L168

Should we ignore it? or we need to solve it?

**Comment by mvashishtha** [comment]
@rosdyana I think we need some more mypy annotations before we can convince mypy that all those classes are not `None`-- right now I think it thinks that some of them are `None`, so the assertions will fail and we'll never reach the code below. Can you add ` # type: ignore[unreachable]` to the end of line 168 (the last assertion)?

File: modin/distributed/dataframe/pandas/partitions.py (hunk 7 above)
Comment (mvashishtha) [reply modin/distributed/dataframe/pandas/partitions.py:27]: @rosdyana yes, could you please do that, following the example here: https://github.com/modin-project/modin/blob/9013f54283eb6776920ee3bf527e208a516d086d/modin/core/dataframe/pandas/dataframe/dataframe.py#L42

File: modin/distributed/dataframe/pandas/partitions.py (hunk 9 above)
Comment (rosdyana) [reply modin/distributed/dataframe/pandas/partitions.py]: Revert back this code.

File: modin/distributed/dataframe/pandas/partitions.py (hunk 8 above)
Comment (rosdyana) [reply modin/distributed/dataframe/pandas/partitions.py]: Done.

File: modin/distributed/dataframe/pandas/partitions.py (hunk 7 above)
Comment (rosdyana) [reply modin/distributed/dataframe/pandas/partitions.py]: @mvashishtha Done.

**Review by pyrito — COMMENTED** [review]
@rosdyana could you rebase on the latest master? There were some changes that broke CI earlier but it should be fixed now.

**Comment by mvashishtha** [comment]
@rosdyana @pyrito no need to rebase on master. I don't think the breaking changes are in the branch for this PR. @rosdyana you should be able to just push an empty commit to get testing working again:

```bash
git commit -sam "Empty commit to trigger CI." --no-verify --allow-empty && git push
```

**Comment by rosdyana** [comment]
Hey, @mvashishtha @pyrito let me know if I need to change anything else. Thanks!

**Comment by vnlitvinov** [comment]
@rosdyana please take a look at CI failure here: https://github.com/modin-project/modin/actions/runs/3289605374/jobs/5432602430

**Review by mvashishtha — APPROVED** [review]

**Review by vnlitvinov — APPROVED** [review]
LGTM, thanks a lot @rosdyana!

**Comment by mvashishtha** [comment]
+1 @rosdyana thank you for your contribution. Please continue reach out on GitHub, slack, or discuss, if you would like to contribute more.
